urllib3/util/ssl_.py

```diff
@@ -323,19 +330,42 @@ def ssl_wrap_socket(sock, keyfile=None, certfile=None, cert_reqs=None,
         # try to load OS default certs; works well on Windows (require Python3.4+)
         context.load_default_certs()
 
+    # Due to a bug in Python's handling of SNI extensions (sends IP as SNI hostname when connecting via IP address)
+    # we shouldn't send things that parse as IP addresses to wrap_socket as a hostname.
+    # If ipaddress is available, use it to filter those out.
+    server_hostname_is_ip = False
+
+    if HAS_IPADDRESS:
+        try:
+            # ipaddress doesn't accept byte strings
+            if isinstance(server_hostname, u''.__class__):
+                server_hostname_uni = server_hostname
+            else:
+                server_hostname_uni = server_hostname.decode('utf-8')
```

Comment:
Should we assume UTF-8? `locale.getpreferredencoding()` might be a safer bet.

Reply:
If the `locale` route is taken we should probably make sure to pass `False` into the function to avoid a `setlocale()`? `setlocale()` isn't thread-safe.